src/kbmod/search/raw_image_eigen.cpp

```diff
@@ -0,0 +1,767 @@
+      throw std::runtime_error("Unable to open FITS file for appending.");
+    }
+
+    // This appends a layer (extension) if the file exists)
+    /* Create the primary array image (32-bit float array) */
+    long dimensions[2];
+    dimensions[0] = width;
+    dimensions[1] = height;
+    fits_create_img(f, FLOAT_IMG, 2 /*naxis*/, dimensions, &status);
+    fits_report_error(stderr, status);
+
+    /* Write the array of floats to the image */
+    fits_write_img(f, TFLOAT, 1, get_npixels(), image.data(), &status);
+    fits_report_error(stderr, status);
+
+    // Save the image time in the header.
+    fits_update_key(f, TDOUBLE, "MJD", &obstime, "[d] Generated Image time", &status);
+    fits_report_error(stderr, status);
+
+    fits_close_file(f, &status);
+    fits_report_error(stderr, status);
+  }
+
+
+  // my implementation here, to test performance for this because no way the
+  // optimization from the original one actually can be worth the messiness
+  const float calc_median(std::vector<float>& values) {
+    std::sort(values.begin(), values.end());
+    const auto len = values.size();
+    return len % 2 == 0 ?
+      (values[len/2-1] + values[len/2]) / 2.0f :
+      values[len/2];
+  }
+
+
+  RawImageEigen create_median_image_eigen2(const std::vector<RawImageEigen>& images) {
+    int num_images = images.size();
+    assert(num_images > 0);
+
+    unsigned height = images[0].get_height();
+    unsigned width = images[0].get_width();
+    RawImageEigen result = RawImageEigen(height, width);
+
+    std::vector<float> pixels;
+    pixels.reserve(num_images);
+    for (unsigned y=0; y<height; ++y){
+      for (unsigned x=0; x<width; ++x){
+        int n_unmasked = 0;
+        for (int i=0; i<num_images; ++i){
+          float pix_val = images[i].get_pixel(x, y);
+          if (pix_val != NO_DATA){
+            pixels.push_back(pix_val);
+          }
+        } // for i
+
+        if (n_unmasked > 0){
+          result.image(x, y) = calc_median(pixels);
+        }
+
+      } // for x
+    } // for y
+
+    return result;
+  }
+
+
+  // default impl
+  RawImageEigen create_median_image_eigen(const std::vector<RawImageEigen>& images) {
+    int num_images = images.size();
+    assert(num_images > 0);
+
+    // if we always use get_pixel why is this neccessary?
+    int width = images[0].get_width();
+    int height = images[0].get_height();
+    for (auto& img : images) {
+      assert(img.get_width() == width and img.get_height() == height);
+    }
+
+    RawImageEigen result = RawImageEigen(width, height);
+
+    std::vector<float> pix_array(num_images);
+    for (int y = 0; y < height; ++y) {
+      for (int x = 0; x < width; ++x) {
+        int num_unmasked = 0;
+        for (int i = 0; i < num_images; ++i) {
+          // Only used the unmasked array.
+          float pix_val = images[i].get_pixel(x, y);
+          if ((pix_val != NO_DATA) && (!std::isnan(pix_val))) { // why are we suddenly checking nans?!
```

Comment:
We should probably check for NaNs when the data is loaded and replace them with NO_DATA or add NaN checks throughout.

Reply:
I don't think it'd be smart to have two different mask values for things. `NaNs` are nice because they are a singleton and we don't have to worry about accidentally overwriting an actual value. Of course, the same issues remain if NaN is baked into the image itself as with `NO_DATA`.